### PR TITLE
fix: meta push split assignments down when recovery

### LINKED ISSUE #4918: bug: streaming jobs (NexMark) didn't recover after cluster restarts
### Describe the bug

As title.

### To Reproduce

1. `./risedev d full`
2. Create NexMark sources and `create materialized view nexmark_q7`
3. `./risedev d full` again to restart cluster
4. Watch Grafana and no throughput

### Expected behavior

Streaming jobs should recover after restart

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tabVersion**
Signed-off-by: tabVersion <tabvision@bupt.icu>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

as title, push the assignment to source executor can help restore state. 

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

fix #4918

### CODE REVIEW

File: src/meta/src/stream/source_manager.rs [hunk 1]
```diff
@@ -716,13 +722,12 @@ where
...
-            tokio_retry::Retry::spawn(FixedInterval::new(Self::SOURCE_RETRY_INTERVAL), || async {
-                let command = command.clone();
-                self.barrier_manager.run_command(command).await
-            })
-            .await
-            .expect("source manager barrier push down failed");
-
+            let (run_command_tx, run_command_rx) = oneshot::channel::<MetaResult<()>>();
+            self.split_diff_tx.send((command, run_command_tx)).unwrap();
+            run_command_rx
```
Comment (yezizp2012) [inline src/meta/src/stream/source_manager.rs:727]: We'd better retry it rather than panic here.

Comment (tabVersion) [reply src/meta/src/stream/source_manager.rs:727]: I think sending to channel is unlikely to fail. we may retry if run command fails.

Comment (yezizp2012) [reply src/meta/src/stream/source_manager.rs:727]: Yes, that's exactly what I meant.

File: src/meta/src/barrier/mod.rs [hunk 2]
```diff
@@ -613,6 +620,14 @@ where
...
                     return;
                 }
+                change_split_req = split_diff_rx.recv() => {
+                    let (command, notify_source_tx) = change_split_req.unwrap();
+                    // enqueue the source split config change command.
+                    self.scheduled_barriers.push(vec![(command,  once(Notifier {
+                        collected: Some(notify_source_tx),
+                        ..Default::default()
+                    }).collect())]).await;
+                }
```
Comment (BugenZhao) [inline src/meta/src/barrier/mod.rs:630]: We may not add a special select arm for split change: `BarrierManager::run` acts like a consumer for the checkpoint queue and does not realize which specific type of command it runs. New commands should be enqueued externally.

Comment (yezizp2012) [reply src/meta/src/barrier/mod.rs:630]: Well, this's a workaround to enqueue split config change command. Previously it's injected by calling `run_command` inside source manager. But for push down split assignments in recovery, the barrier manager will also need a reference of source manager, which will lead to circular references.
🤔, is there any other way we can cover these two scenarios and not lead to circular references?

File: src/meta/src/rpc/server.rs [hunk 3]
```diff
@@ -483,7 +486,7 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
             ClusterManager::start_heartbeat_checker(cluster_manager, Duration::from_secs(1)).await,
         );
 
-        sub_tasks.push(GlobalBarrierManager::start(barrier_manager).await);
+        sub_tasks.push(GlobalBarrierManager::start(barrier_manager, split_diff_rx).await);
```
Comment (tabVersion) [inline src/meta/src/rpc/server.rs:489]: @yezizp2012  a little off the topic, why GlobalBarrierManager does not start in test mode (`#[cfg(not(test))]`).

Comment (yezizp2012) [reply src/meta/src/rpc/server.rs:489]: Well, `rpc_serve_with_store` is only used in `server.rs` in the test mode and few things are tested. Actually all sub tasks can be configured to not start if we don't care about it and need it in the test mode.

File: src/stream/src/executor/source/source_executor.rs [hunk 4]
```diff
@@ -245,7 +245,7 @@ impl<S: StateStore> SourceExecutor<S> {
             .stack_trace("source_recv_first_barrier")
             .await
             .unwrap();
-
+        println!("{:?} barrier {:?}", self.ctx.id, barrier);
```
Comment (tabVersion) [inline src/stream/src/executor/source/source_executor.rs:248]: got None mutation here

File: src/meta/src/barrier/mod.rs [hunk 5]
```diff
@@ -568,17 +573,24 @@ where
...
 
         (join_handle, shutdown_tx)
     }
 
     /// Start an infinite loop to take scheduled barriers and send them.
-    async fn run(&self, mut shutdown_rx: Receiver<()>) {
+    async fn run(
+        &self,
+        mut shutdown_rx: Receiver<()>,
+        mut split_diff_rx: UnboundedReceiver<(Command, Sender<MetaResult<()>>)>,
```
Comment (shanicky) [inline src/meta/src/barrier/mod.rs:592]: maybe we can use a better name?

File: src/meta/src/barrier/mod.rs [hunk 6]
```diff
@@ -613,6 +625,16 @@ where
                     tracing::info!("Barrier manager is stopped");
                     return;
                 }
+                // FIXME: recovery needs split assignments on all actors, this requires barrier manager holds a ref of source manager.
+                // Here is a walk around for cycle dep: source manager sends diff via a channel and consume here.
+                change_split_req = split_diff_rx.recv() => {
+                    let (command, notify_source_tx) = change_split_req.unwrap();
+                    // enqueue the source split config change command.
+                    self.scheduled_barriers.push(vec![(command,  once(Notifier {
```
Comment (shanicky) [inline src/meta/src/barrier/mod.rs:633]: space before once?

**Review by shanicky — APPROVED** [review]
rest generally LGTM

**Review by yezizp2012 — COMMENTED** [review]
Generally LGTM

**Pull request merged**